app/src/main/java/net/squanchy/schedule/tracksfilter/widget/FilterChipView.kt

```diff
@@ -0,0 +1,239 @@
+        checkedTextColorLight = a.getColor(R.styleable.FilterChipView_checkedTextColorLight, Color.MAGENTA)
+        uncheckedFallbackTextColor = a.getColor(R.styleable.FilterChipView_uncheckedFallbackTextColor, Color.GREEN)
+
+        baseColor = a.getColor(R.styleable.FilterChipView_color, ContextCompat.getColor(context, R.color.chip_default_background_tint))
+
+        strokeWidth = a.getDimensionPixelSize(R.styleable.FilterChipView_strokeWidth, 0)
+
+        backgroundDrawable = ResourcesCompat.getDrawable(resources, R.drawable.chip_background, context.theme) as GradientDrawable
+        super.setBackground(backgroundDrawable)
+
+        backgroundAlpha = CheckableValue(
+            checkedValue = a.getFloat(R.styleable.FilterChipView_checkedAlpha, DEFAULT_CHECKED_ALPHA),
+            uncheckedValue = a.getFloat(R.styleable.FilterChipView_uncheckedAlpha, DEFAULT_UNCHECKED_ALPHA)
+        )
+
+        a.recycle()
+
+        isClickable = true
+        isFocusable = true
+
+        super.setOnClickListener { toggle() }
+        isInitialized = true
+
+        updateColors(baseColor)
+    }
+
+    var color: Int
+        @ColorInt
+        get() = baseColor
+        set(@ColorInt color) = updateColors(color)
+
+    private val parentBackgroundColor
+        get() = ASSUMED_PARENT_BACKGROUND_COLOR // TODO allow setting a proper parent color via code if necessary
+
+    private fun updateColors(@ColorInt baseColor: Int) {
+        this.baseColor = baseColor
+        backgroundColor = CheckableValue(checkedValue = baseColor, uncheckedValue = Color.TRANSPARENT)
+        strokeColor = CheckableValue(checkedValue = baseColor, uncheckedValue = baseColor)
+
+        val checkedTextColor = baseColor.pickBestTextColorByContrast(checkedTextColorLight, checkedTextColorDark)
+        val contrastSafeBaseColor = baseColor.darkenToEnsureTextContrasts(parentBackgroundColor, uncheckedFallbackTextColor)
+        val uncheckedTextColor = parentBackgroundColor.pickBestTextColorByContrast(baseColor, contrastSafeBaseColor)
+        textColor = CheckableValue(checkedValue = checkedTextColor, uncheckedValue = uncheckedTextColor)
+
+        redraw()
+    }
+
+    override fun isChecked(): Boolean = isChecked
+
+    override fun toggle() {
+        setChecked(!isChecked)
+    }
+
+    override fun setChecked(checked: Boolean) {
+        if (checked == isChecked) return
+        isChecked = checked
+
+        redraw()
+
+        // Avoid infinite recursions if checked is set from a listener
+        if (isBroadcastingCheckedChange) {
+            return
+        }
+
+        isBroadcastingCheckedChange = true
+        onCheckedChangeListener?.invoke(this, checked)
+        isBroadcastingCheckedChange = false
+    }
+
+    private fun redraw() {
+        updateBackgroundDrawable()
+        updateTextColor()
+        invalidate()
+    }
+
+    private fun updateBackgroundDrawable() {
+        with(backgroundDrawable) {
```

Comment:
Maybe apply instead?

Reply:
/shrug yeah not sure why I used with 💃

Reply:
966cfc707a20c4f1b5ed29606ecaf5bbd6ca9053